tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -139,8 +183,332 @@ impl ServerKey {
+                            &mut *current_block,
+                            &*previous_block,
+                            &lut_carry_propagation_sum,
+                        );
+                    }
+                });
+        }
+
+        // Down-Sweep phase
+        let mut buffer = Vec::with_capacity(num_blocks / 2);
+        self.key
+            .create_trivial_assign(&mut carry_out[num_blocks - 1], 0);
+        for i in (0..num_steps).rev() {
+            let two_pow_i_plus_1 = 2usize.checked_pow((i + 1) as u32).unwrap();
+            let two_pow_i = 2usize.checked_pow(i as u32).unwrap();
+
+            (0..num_blocks)
+                .into_par_iter()
+                .step_by(two_pow_i_plus_1)
+                .map(|k| {
+                    // Since our carry_propagation LUT ie sum function
+                    // is not commutative we have to reverse operands
+                    self.key.unchecked_apply_lookup_table_bivariate(
+                        &carry_out[k + two_pow_i - 1],
+                        &carry_out[k + two_pow_i_plus_1 - 1],
+                        &lut_carry_propagation_sum,
+                    )
+                })
+                .collect_into_vec(&mut buffer);
+
+            let mut drainer = buffer.drain(..);
+            for k in (0..num_blocks).step_by(two_pow_i_plus_1) {
+                let b = drainer.next().unwrap();
+                carry_out.swap(k + two_pow_i - 1, k + two_pow_i_plus_1 - 1);
+                carry_out[k + two_pow_i_plus_1 - 1] = b;
+            }
+            drop(drainer);
+            assert!(buffer.is_empty());
+        }
+
+        // The first step of the Down-Sweep phase sets the
+        // first block to 0, so no need to re-do it
+        lhs.blocks
+            .par_iter_mut()
+            .zip(carry_out.par_iter())
+            .for_each(|(block, carry_in)| {
+                self.key.unchecked_add_assign(block, carry_in);
+                self.key.message_extract_assign(block);
+            });
+    }
+
+    /// Initialization function for parallal carryless sum
+    ///
+    /// This function adds rhs into lhs
+    ///
+    /// It returns the array that tells whether the
+    /// block resulting from the sum will propagate or generate
+    /// a carry. A Prefix sum / cumulative sum then needs to be done
+    /// to get the final carry that the block will output
+    fn add_and_generate_init_carry_array<PBSOrder: PBSOrderMarker>(
+        &self,
+        lhs: &mut RadixCiphertext<PBSOrder>,
+        rhs: &RadixCiphertext<PBSOrder>,
+        add_extra_one: AddExtraOne,
+    ) -> Vec<crate::shortint::CiphertextBase<PBSOrder>> {
+        let modulus = self.key.message_modulus.0 as u64;
+
+        // This used for the first pair of blocks
+        // as this pair can only generate or not but never propagate
+        let lut_does_block_generate_carry = self.key.generate_accumulator(|x| {
+            if x >= modulus - 1 {
```

Comment:
is this correct the previous code was x >= modulus for generated

Reply:
:hot_face: trying to be too fast